lib/pegboard/manager/src/utils.rs

```diff
@@ -43,3 +83,167 @@ pub fn get_s3_path_stub(url: &Url, with_uuid: bool) -> Result<PathBuf> {
+		.connect(db_url)
+		.await
+		.map_err(Into::into)
+}
+
+// TODO: Replace with migrations
+pub async fn init_sqlite_schema(pool: &SqlitePool) -> Result<()> {
+	// Attempt to use an existing connection
+	let mut conn = if let Some(conn) = pool.try_acquire() {
+		conn
+	} else {
+		// Create a new connection
+		pool.acquire().await?
+	};
+
+	sqlx::query(indoc!(
+		"
+		CREATE TABLE IF NOT EXISTS events (
+			data BLOB NOT NULL,
+			create_ts INTEGER NOT NULL
+		)
+		",
+	))
+	.execute(&mut *conn)
+	.await?;
+
+	// sqlx::query(indoc!(
+	// 	"
+	// 	CREATE TABLE IF NOT EXISTS commands (
+
+	// 	)
+	// 	",
+	// ))
+	// .execute(&mut *conn)
+	// .await?;
+
+	sqlx::query(indoc!(
+		"
+		CREATE TABLE IF NOT EXISTS containers (
+			container_id TEXT PRIMARY KEY, -- UUID
+			create_ts INTEGER NOT NULL,
+			stop_ts INTEGER,
+			exit_ts INTEGER,
+			exit_code INTEGER
+		)
+		",
+	))
+	.execute(&mut *conn)
+	.await?;
+
+	Ok(())
+}
+
+/// Executes queries and explicitly handles retry errors.
+pub async fn query<'a, F, Fut, T>(mut cb: F) -> Result<T>
+where
+	F: FnMut() -> Fut,
+	Fut: std::future::Future<Output = std::result::Result<T, sqlx::Error>> + 'a,
+	T: 'a,
+{
+	let mut i = 0;
+
+	loop {
+		match cb().await {
+			std::result::Result::Ok(x) => return Ok(x),
+			std::result::Result::Err(err) => {
+				use sqlx::Error::*;
+
+				if i > MAX_QUERY_RETRIES {
+					bail!("max sql retries: {err}");
+				}
+				i += 1;
+
+				match &err {
+					// Retry transaction errors immediately
+					Database(db_err)
+						if db_err
+							.message()
+							.contains("TransactionRetryWithProtoRefreshError") =>
+					{
+						tracing::info!(message=%db_err.message(), "transaction retry");
+						tokio::time::sleep(TXN_RETRY).await;
+					}
+					// Retry internal errors with a backoff
+					Database(_) | Io(_) | Tls(_) | Protocol(_) | PoolTimedOut | PoolClosed
+					| WorkerCrashed => {
+						tracing::info!(?err, "query retry");
+						tokio::time::sleep(QUERY_RETRY).await;
+					}
+					// Throw error
+					_ => return Err(err.into()),
+				}
+			}
+		}
+	}
+}
+
+pub fn now() -> i64 {
+	time::SystemTime::now()
+		.duration_since(time::UNIX_EPOCH)
+		.unwrap_or_else(|err| unreachable!("time is broken: {}", err))
+		.as_millis()
+		.try_into()
+		.expect("now doesn't fit in i64")
+}
+
+/// Creates an async file watcher.
+fn async_watcher() -> Result<(RecommendedWatcher, Receiver<notify::Result<Event>>)> {
+	let (tx, rx) = channel(1);
+
+	// Automatically select the best implementation for your platform.
+	let watcher = RecommendedWatcher::new(
+		move |res| {
+			let tx = tx.clone();
+
+			// Send event. We ignore the result because the watcher and channel are dropped after the first
+			// `Create` event is received in wait_for_creation
+			let _ = tx.blocking_send(res);
+		},
+		Config::default().with_poll_interval(Duration::from_secs(2)),
+	)?;
+
+	Ok((watcher, rx))
+}
+
+pub async fn wait_for_creation<P: AsRef<Path>>(path: P) -> Result<()> {
```

Comment:
make sure the systemd service has enough watchers configured. write the calculation in the comments (i.e. `supports X containers * Y watchers per container`) and comment in this func to update the systemd watcher count too